kornia/enhance/equalization.py

```diff
@@ -0,0 +1,306 @@
+                if residual:
+                    hist[0:int(residual)] += 1
+
+    lut_scale = (num_bins - 1) / pixels
+    luts = torch.cumsum(histos, 1) * lut_scale
+    luts = luts.clamp(0, num_bins - 1).floor()  # to get the same values as converting to int maintaining the type
+    luts = luts.view(([*tiles_x_im.shape[0:4]] + [num_bins]))
+    return luts
+
+
+def _map_luts(interp_tiles: torch.Tensor, luts: torch.Tensor) -> torch.Tensor:
+    r"""Assign the required luts to each tile.
+
+    Args:
+        interp_tiles (torch.Tensor): set of interpolation tiles. (B, 2GH, 2GW, C, TH/2, TW/2)
+        luts (torch.Tensor): luts for each one of the original tiles. (B, GH, GW, C, 256)
+
+    Returns:
+        torch.Tensor: mapped luts (B, 2GH, 2GW, 4, C, 256)
+
+    """
+    num_imgs: int  # number of batched images
+    gh: int  # 2x the number of tiles used to compute the histograms
+    gw: int
+    c: int  # number of channels
+    th: int  # /2 the sizes of the tiles used to compute the histograms
+    tw: int
+    num_imgs, gh, gw, c, th, tw = interp_tiles.shape
+
+    # precompute idxs for non corner regions (doing it in cpu seems sligthly faster)
+    j_idxs = torch.ones(gh - 2, 4, dtype=torch.long) * torch.arange(1, gh - 1).reshape(gh - 2, 1)
+    i_idxs = torch.ones(gw - 2, 4, dtype=torch.long) * torch.arange(1, gw - 1).reshape(gw - 2, 1)
+    j_idxs = j_idxs // 2 + j_idxs % 2
+    j_idxs[:, [0, 1]] -= 1
+    i_idxs = i_idxs // 2 + i_idxs % 2
+    i_idxs[:, [0, 2]] -= 1
+
+    # selection of luts to interpolate each patch
+    # create a tensor with dims: interp_patches height and width x 4 x num channels x bins in the histograms
+    # the tensor is init to -1 to denote non init hists
+    luts_x_interp_tiles: torch.Tensor = -torch.ones(
+        num_imgs, gh, gw, 4, c, luts.shape[-1], device=interp_tiles.device)  # B x GH x GW x 4 x C x 256
+    # corner regions
+    luts_x_interp_tiles[:, 0::gh - 1, 0::gw - 1, 0] = luts[:, 0::max(gh // 2 - 1, 1), 0::max(gw // 2 - 1, 1)]
+    # border region (h)
+    luts_x_interp_tiles[:, 1:-1, 0::gw - 1, 0] = luts[:, j_idxs[:, 0], 0::max(gw // 2 - 1, 1)]
+    luts_x_interp_tiles[:, 1:-1, 0::gw - 1, 1] = luts[:, j_idxs[:, 2], 0::max(gw // 2 - 1, 1)]
+    # border region (w)
+    luts_x_interp_tiles[:, 0::gh - 1, 1:-1, 0] = luts[:, 0::max(gh // 2 - 1, 1), i_idxs[:, 0]]
+    luts_x_interp_tiles[:, 0::gh - 1, 1:-1, 1] = luts[:, 0::max(gh // 2 - 1, 1), i_idxs[:, 1]]
+    # internal region
+    luts_x_interp_tiles[:, 1:-1, 1:-1, :] = luts[
+        :, j_idxs.repeat(max(gh - 2, 1), 1, 1).permute(1, 0, 2), i_idxs.repeat(max(gw - 2, 1), 1, 1)]
+
+    return luts_x_interp_tiles
+
+
+def _compute_equalized_tiles(interp_tiles: torch.Tensor, luts: torch.Tensor) -> torch.Tensor:
+    r"""Equalize the tiles.
+
+    Args:
+        interp_tiles (torch.Tensor): set of interpolation tiles. (B, 2GH, 2GW, C, TH/2, TW/2)
+        luts (torch.Tensor): luts for each one of the original tiles. (B, GH, GW, C, 256)
```

Comment:
is 256 fixed ? can this be dynamic ?

Reply:
Now is fixed, in the next version will be dynamic.